--- extsrc/src/score.c.orig	2011-11-25 11:23:08.316843145 -0800
+++ extsrc/src/score.c	2011-11-25 11:22:01.428641100 -0800
@@ -467,6 +467,14 @@
 		highscore_read(scores, N_ELEMENTS(scores));
 		highscore_add(&entry, scores, N_ELEMENTS(scores));
 		highscore_write(scores, N_ELEMENTS(scores));
+		android_score_start();
+ 		android_score_detail("name", entry.who);
+		android_score_detail("race", entry.p_r);
+		android_score_detail("class", entry.p_c);
+		android_score_detail("how_died", entry.how);
+		android_score_detail("gold", entry.gold);
+		android_score_detail("turns", entry.turns);
+		android_score_submit(entry.pts, entry.max_lev);
 	}
 
 	/* Success */
