diff -Nru --strip-trailing-cr npp-6.1.0beta2-src/src/death.c npp-6.1.0beta2-src-android/src/death.c
--- npp-6.1.0beta2-src/src/death.c	2012-09-26 18:05:50.000000000 -0700
+++ npp-6.1.0beta2-src-android/src/death.c	2013-01-28 19:55:29.904032000 -0800
@@ -498,7 +498,8 @@
 
 		if (c.key == ESCAPE || cursor == (menu->count -1))
 		{
-			if (get_check("Do you want to quit? "))
+			char opts[4],ch;sprintf(opts,"yn%c",KTRL('X'));ch=get_char("Do you want to quit? ", opts, 3, 'n');
+			if (ch == 'y' || ch == KTRL('X'))
 				break;
 		}
 		else if (c.type == EVT_SELECT && death_actions[cursor].action)
diff -Nru --strip-trailing-cr npp-6.1.0beta2-src/src/h-basic.h npp-6.1.0beta2-src-android/src/h-basic.h
--- npp-6.1.0beta2-src/src/h-basic.h	2009-11-16 02:13:14.000000000 -0800
+++ npp-6.1.0beta2-src-android/src/h-basic.h	2013-01-26 21:42:14.690702000 -0800
@@ -75,7 +75,7 @@
  * Basically, SET_UID should *only* be set for "Unix" machines.
  */
 #if !defined(MACH_O_CARBON) && !defined(WINDOWS) && \
-    !defined(RISCOS) && !defined(GAMEBOY) && !defined(NDS)
+    !defined(RISCOS) && !defined(GAMEBOY) && !defined(NDS) && !defined(USE_AND)
 # define SET_UID
 
 /* Without autoconf, turn on some things */
