diff -Nru --strip-trailing-cr angband-nightly/lib/file/news.txt angband-nightly-android/lib/file/news.txt
--- angband-nightly/lib/file/news.txt	2011-02-25 14:06:01.000000000 -0800
+++ angband-nightly-android/lib/file/news.txt	2011-02-25 13:56:58.000000000 -0800
@@ -6,7 +6,7 @@
 {yellow}                 /_/   \_\_| |_|\__, |_.__/ \__,_|_| |_|\__,_|{/}
 {yellow}                                |___/                         {/}  
 
-                              Version $VERSION
+                   Version $VERSION
                 Bug reports to <http://angband.oook.cz/forum/>
 
    For resources and links to places you can talk about the game, please see:
diff -Nru --strip-trailing-cr angband-nightly/src/buildid.h angband-nightly-android/src/buildid.h
--- angband-nightly/src/buildid.h	2011-02-25 12:12:23.000000000 -0800
+++ angband-nightly-android/src/buildid.h	2011-02-25 11:35:44.000000000 -0800
@@ -8,7 +8,7 @@
 # define XSTR(x) STR(x)
 # define VERSION_STRING XSTR(BUILD_ID)
 #else
-# define VERSION_STRING "3.3.0"
+# define VERSION_STRING "3.3.0 built " __DATE__ " " __TIME__
 #endif
 
 extern const char *buildid;
diff -Nru --strip-trailing-cr angband-nightly/src/h-basic.h angband-nightly-android/src/h-basic.h
--- angband-nightly/src/h-basic.h	2011-02-25 12:12:23.000000000 -0800
+++ angband-nightly-android/src/h-basic.h	2011-02-24 23:07:59.000000000 -0800
@@ -64,7 +64,7 @@
  * Basically, SET_UID should *only* be set for "Unix" machines.
  */
 #if !defined(MACH_O_CARBON) && !defined(WINDOWS) && \
-		!defined(GAMEBOY) && !defined(NDS)
+		!defined(GAMEBOY) && !defined(NDS) && !defined(USE_AND)
 # define SET_UID
 
 /* Without autoconf, turn on some things */
